src/services/shopping_list_service.py

```diff
@@ -8,58 +8,92 @@
-
+            meal_names = self._get_meal_names(user_plan)
             for meal_name in meal_names:
-                new_ingredients: set[str] = self._get_ingredients_for_meals(user_id, [meal_name])
-                logging.info(f"Ingredients for {meal_name} on {current_date}: {new_ingredients}")
-
+                new_ingredients = await self._get_ingredients_for_meals(
+                    user_id, [meal_name]
+                )
+                logging.info(
+                    f"Ingredients for {meal_name} on {current_date}: {new_ingredients}"
+                )
                 ingredients.update(new_ingredients)
 
         if not ingredients:
-            logging.warning(f"No ingredients found for user {user_id} in the date range {start_date} to {end_date}.")
+            logging.warning(
+                f"No ingredients found for user {user_id} in the date range "
+                f"{start_date} to {end_date}."
+            )
             return set()
 
         logging.info(f"Final ingredients set: {ingredients}")
         return ingredients
 
-    def _get_meal_names(self, user_plan: dict[str, int | date | str]) -> Generator[str, None, None]:
-        for meal in ['breakfast', 'lunch', 'dinner', 'dessert']:
-            if (meal_info := user_plan.get(meal)):
-                if isinstance(meal_info, str) and (meal_name := self._extract_meal_name(meal_info)):
+    def _get_meal_names(
+        self, user_plan: dict[str, int | date | str]
+    ) -> Generator[str, None, None]:
+        """Extract meal names from user plan."""
+        for meal in ["breakfast", "lunch", "dinner", "dessert"]:
+            if meal_info := user_plan.get(meal):
+                if isinstance(meal_info, str) and (
+                    meal_name := self._extract_meal_name(meal_info)
+                ):
                     yield meal_name
 
     def _extract_meal_name(self, meal_info: str) -> str | None:
-        if '(ID:' in meal_info:
-            return meal_info.split('(ID:')[0].strip()
+        """Extract meal name from meal info string."""
+        if "(ID:" in meal_info:
+            return meal_info.split("(ID:")[0].strip()
         return meal_info
 
-    def _get_ingredients_for_meals(self, user_id: int, meal_names: list[str]) -> set[str]:
+    async def _get_ingredients_for_meals(
+        self, user_id: int, meal_names: list[str]
+    ) -> set[str]:
+        """Get ingredients for specified meals."""
         ingredients: set[str] = set()
         for meal_name in meal_names:
-            if (recipe := self.recipe_manager.get_recipe_by_name(user_id, meal_name)):
-                logging.info(f"Recipe for {meal_name}: {recipe}")
-                if 'ingredients' in recipe:
-                    ingredients.update(recipe['ingredients'])
-        return ingredients
\ No newline at end of file
+            recipe = await self.recipe_manager.get_recipe_by_name(user_id, meal_name)

```

Comment:
to truly leverage the async nature of these ops:
```suggestion
        recipes = await asyncio.gather(*[
            self._get_recipe_with_sem(sem, user_id, name)
            for name in meal_names
        ])
```